pkg/deployer/helm/ensure.go

```diff
@@ -261,32 +289,30 @@ func (h *Helm) cleanupOrphanedResources(ctx context.Context, kubeClient client.C
 		wg      sync.WaitGroup
 	)
 	for _, ref := range oldObjects {
-		uObj := unstructured.Unstructured{
-			Object: map[string]interface{}{
-				"apiVersion": ref.APIVersion,
-				"kind":       ref.Kind,
-				"metadata": map[string]interface{}{
-					"name":      ref.Name,
-					"namespace": ref.Namespace,
-				},
-			},
-		}
-		if err := kubeClient.Get(ctx, kutil.ObjectKey(ref.Name, ref.Namespace), &uObj); err != nil {
+		obj := kutil.ObjectFromTypedObjectReference(&ref)
+		if err := kubeClient.Get(ctx, kutil.ObjectKey(ref.Name, ref.Namespace), obj); err != nil {
 			if apierrors.IsNotFound(err) {
 				continue
 			}
-			return fmt.Errorf("unable to get object %s %s: %w", uObj.GroupVersionKind().String(), uObj.GetName(), err)
+			return fmt.Errorf("unable to get object %s %s: %w", obj.GroupVersionKind().String(), obj.GetName(), err)
 		}
 
-		if !containsUnstructuredObject(&uObj, currentObjects) {
+		if !containsUnstructuredObject(obj, currentObjects) {
 			wg.Add(1)
-			go func(obj unstructured.Unstructured) {
+			go func(obj *unstructured.Unstructured) {
 				defer wg.Done()
-				if err := kubeClient.Delete(ctx, &obj); err != nil {
+				if err := kubeClient.Delete(ctx, obj); err != nil {
 					allErrs = append(allErrs, fmt.Errorf("unable to delete %s %s/%s: %w", obj.GroupVersionKind().String(), obj.GetName(), obj.GetNamespace(), err))
 				}
-				// todo: wait for deletion
-			}(uObj)
+
+				pollCtx, cancel := context.WithTimeout(ctx, 60*time.Second)
+				defer cancel()
+				delCondFunc := kutil.GenerateDeleteObjectConditionFunc(ctx, kubeClient, obj)
+				err := wait.PollImmediateUntil(5*time.Second, delCondFunc, pollCtx.Done())
+				if err != nil {
+					allErrs = append(allErrs, err)
+				}
+			}(obj)
```

Comment:
can you add a comment why this is needed.
It looks a lot like the code in the manifest deployer. Maybe we can refactor that into one function/struct that is used by both.

Reply:
yes, definitely something that can goes into kubernetes utils